Yet more `parse_tt` improvements

nnethercote opened the pull request:
Including lots of comment improvements, and an overhaul of how `matches` work that gives big speedups.

r? @petrochenkov

nnethercote commented:
Some `check full` instruction count results.

Benchmark & Profile | Scenario | % Change | Significance Factor?
-- | -- | -- | --
quote-stress check | full | -17.38% | 86.90x
async-std-1.10.0 check | full | -16.64% | 83.18x
time-macros-0.2.3 check | full | -10.55% | 52.74x
yansi-0.5.0 check | full | -9.62% | 48.10x
deep-vector check | full | -2.85% | 14.26x
num-derive-0.3.3 check | full | -2.51% | 12.56x
pest_generator-2.1.3 check | full | -2.45% | 12.27x
inotify-0.10.0 check | full | -2.30% | 11.49x
scroll_derive-0.11.0 check | full | -2.28% | 11.41x
ctor-0.1.21 check | full | -2.25% | 11.26x
mockall_derive-0.11.0 check | full | -1.98% | 9.90x
futures-macro-0.3.19 check | full | -1.89% | 9.46x
html5ever check | full | -0.85% | 4.27x
token-stream-stress check | full | 0.66% | 3.32x
syn-1.0.89 check | full | -0.48% | 2.38x
hyper-0.14.18 check | full | -0.41% | 2.07x

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 89325098160f8ee473b0ebfc1b05c323254a1f19 with merge 87e9e2c9089a81be4a5b32d55832388b56a3e885...
<!-- homu: {"type":"TryBuildStarted","head_sha":"89325098160f8ee473b0ebfc1b05c323254a1f19","merge_sha":"87e9e2c9089a81be4a5b32d55832388b56a3e885"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5733548475?check_suite_focus=true)
Build commit: 87e9e2c9089a81be4a5b32d55832388b56a3e885 (`87e9e2c9089a81be4a5b32d55832388b56a3e885`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5733548475?check_suite_focus=true"},"merge_sha":"87e9e2c9089a81be4a5b32d55832388b56a3e885"} -->

rust-timer commented:
Queued 87e9e2c9089a81be4a5b32d55832388b56a3e885 with parent ad5b6f6b72091261ac421d1331b2350f628d1fcb, future [comparison URL](https://perf.rust-lang.org/compare.html?start=ad5b6f6b72091261ac421d1331b2350f628d1fcb&end=87e9e2c9089a81be4a5b32d55832388b56a3e885).

petrochenkov commented on compiler/rustc_expand/src/mbe/macro_parser.rs:
```diff
@@ -254,24 +254,24 @@ crate enum ParseResult<T> {
-        count
-            + match elt {
-                TokenTree::Delimited(_, delim) => count_names(delim.inner_tts()),
+/// Count how many metavars are named `matcher`.
```
Looks like some words are missing in this sentence.

petrochenkov commented on compiler/rustc_expand/src/mbe/macro_parser.rs:
```diff
@@ -254,24 +254,24 @@ crate enum ParseResult<T> {
-            + match elt {
-                TokenTree::Delimited(_, delim) => count_names(delim.inner_tts()),
+/// Count how many metavars are named `matcher`.
+pub(super) fn count_metavars(matcher: &[TokenTree]) -> usize {
```
```suggestion
pub(super) fn count_metavar_decls(matcher: &[TokenTree]) -> usize {
```

petrochenkov commented on compiler/rustc_expand/src/mbe/macro_parser.rs:
```diff
@@ -254,24 +254,24 @@ crate enum ParseResult<T> {
+                // RHS meta-variable expressions eventually end-up here. `0` is returned to inform that
+                // no meta-variable was found, because "meta-variables" != "meta-variable
+                // expressions".
                 TokenTree::MetaVarExpr(..) => 0,
```
This should be unreachable `count_metavars` should never be called when parsing RHS (`parsing_patterns == false`) in the first place, it's 0 there by definition.

nnethercote replied:
I originally thought that too, but if you make it unreachable these tests fail:
```
failures:
    [ui] ui/macros/rfc-3086-metavar-expr/count-and-length-are-distinct.rs
    [ui] ui/macros/rfc-3086-metavar-expr/feature-gate-macro_metavar_expr.rs
    [ui] ui/macros/rfc-3086-metavar-expr/macro-expansion.rs
    [ui] ui/macros/rfc-3086-metavar-expr/out-of-bounds-arguments.rs
```
There may be something to clean up here, but I'd prefer to do it in a follow-up. This change was just meant to give this function a nicer name and slightly cleaner code, it wasn't doing anything about where/how it is called.

petrochenkov replied:
Ok.

petrochenkov commented on compiler/rustc_expand/src/mbe/macro_parser.rs:
```diff
@@ -40,28 +40,28 @@
 //! Remaining input: a a a b
 //! cur: [a · $( a )* a b]
-//! Descend/Skip (first item).
+//! Descend/Skip (first mp).
```
```rust
//! Descend/Skip (first position).
```
I understand using "mp"s in variable names, but in comments I think we can afford being more wordy.

petrochenkov commented on compiler/rustc_expand/src/mbe/macro_parser.rs:
```diff
@@ -89,15 +89,13 @@ use std::borrow::Cow;
-    elts: &'tt [TokenTree],
-    /// The position of the "dot" in `elts` at the time we descended.
+struct MatcherPosFrame<'tt> {
+    /// The "parent" matcher that we have descending from.
```
```rust
    /// The "parent" matcher that we have been descending from.
```
Is this a grammatical sentence?

nnethercote replied:
Not really! :)  "we have descended from" is better.

petrochenkov commented on compiler/rustc_expand/src/mbe/macro_parser.rs:
```diff
@@ -402,11 +419,21 @@ pub struct TtParser<'tt> {
+
+    /// Pre-allocate an empty match array, so it can be cloned cheaply for macros with many rules
+    /// that have no metavars.
+    empty_matches: Lrc<NamedMatchVec>,
```
Nit: perhaps it's better to move this optimization to a separate commit, the "Overhaul how matches are recorded" is already complex without it.

nnethercote replied:
Ok.

rust-timer commented:
Finished benchmarking commit (87e9e2c9089a81be4a5b32d55832388b56a3e885): [comparison url](https://perf.rust-lang.org/compare.html?start=ad5b6f6b72091261ac421d1331b2350f628d1fcb&end=87e9e2c9089a81be4a5b32d55832388b56a3e885).

**Summary**: This benchmark run shows 45 relevant improvements 🎉 to instruction counts.
- Arithmetic mean of relevant improvements: -1.4%
- Arithmetic mean of all relevant changes: -1.4%
- Largest improvement in instruction counts: -6.2% on `incr-unchanged` builds of `deep-vector check`

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

nnethercote commented:
I have updated the code to address comments. Most of the changes are in the relevant commits, but I split out the `empty_matches` change into a new commit.

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 6b0a16ab1ad689679d5b2376f842ef0606bd81c7 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 6b0a16ab1ad689679d5b2376f842ef0606bd81c7 -->
<!-- homu: {"type":"Approved","sha":"6b0a16ab1ad689679d5b2376f842ef0606bd81c7","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 6b0a16ab1ad689679d5b2376f842ef0606bd81c7 with merge c5cf08d37b85f953b132951e868df5b924250fdc...
<!-- homu: {"type":"BuildStarted","head_sha":"6b0a16ab1ad689679d5b2376f842ef0606bd81c7","merge_sha":"c5cf08d37b85f953b132951e868df5b924250fdc"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5761813899?check_suite_focus=true)
Approved by: petrochenkov
Pushing c5cf08d37b85f953b132951e868df5b924250fdc to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5761813899?check_suite_focus=true"},"merge_sha":"c5cf08d37b85f953b132951e868df5b924250fdc"} -->
